bgpd/bgpd.c

```diff
@@ -4023,6 +4023,64 @@ static void peer_flag_modify_action(struct peer *peer, uint32_t flag)
 		bgp_session_reset(peer);
 }
 
+/* Enable global administrative shutdown of all peers of BGP instance */
+void bgp_shutdown_enable(struct bgp *bgp, char *msg)
+{
+	struct peer *peer;
+	struct listnode *node;
+
+	/* do nothing if already shut down */
+	if (CHECK_FLAG(bgp->flags, BGP_FLAG_SHUTDOWN))
+		return;
+
+	/* informational log message */
+	zlog_info("Enabled administrative shutdown on BGP instance AS %u",
+		  bgp->as);
+
+	/* iterate through peers of BGP instance */
+	for (ALL_LIST_ELEMENTS_RO(bgp->peer, node, peer)) {
```

Comment:
does it work even for bgp instances under vrfs?

Reply:
I was advised to copy the functionality of the existing neighbour shutdown, so i have not yet concerned myself with expanding the scope to that.